Clarify behavior of template scripts (#5472)
nextflow-io/nextflow: docs/process.md

### Template

> Process scripts can be externalised to **template** files, which allows them to be reused across different processes and tested independently from the pipeline execution.

Reviewer: ```suggestion
Process scripts can be externalized to **template** files, which allows them to be reused across different processes and tested independently from the pipeline execution.
```

}
```

> By default, Nextflow looks for `my_script.sh` in the `templates` directory located alongside the Nextflow script in which the process is defined. An absolute path can be used to specify a different location, but this practice is discouraged because it hinders pipeline portability.

Reviewer: ```suggestion
By default, in the above example, Nextflow looks for `my_script.sh` in the `templates` directory located alongside the Nextflow script in which the process is defined. An absolute path can be used to specify a different location. However, this practice is discouraged because it hinders pipeline portability.
```

echo "process completed"
```

Variables prefixed with the dollar character (`$`) are interpreted as Nextflow variables when the template script is executed by Nextflow, and Bash variables when executed directly. For example, the above script can be executed from the command line by providing each input as an environment variable:

```bash
STR='foo' bash templates/my_script.sh
```

> The following caveats should be kept in mind:

Reviewer: ```suggestion
The following caveats should be considered:
```

- In practice, templates should be used only for Bash scripts. Languages like Python and R do not prefix variables with `$`, so templates written in these languages can't be executed separately from the pipeline.

- Variables escaped with `\$` will be interpreted as Bash variables when executed by Nextflow, but will not be interpreted as variables at all when executed from the command line. As a result, this practice should be avoided to ensure that the template script behaves consistently in both cases.

> - If a Nextflow variable is missing when the template is executed by Nextflow, the pipeline will fail, even if the missing variable is "commented out" in the template. This happens because the template rendering is not aware of the language in which the template is written.

Reviewer: ```suggestion
- Template rendering is not aware of the language in which the template is written. The pipeline will fail if a Nextflow variable is missing when the template is executed by Nextflow. For example, a pipeline will fail if the missing variable is "commented out" in the template.0
```

:::{tip}

> Due to the caveats listed above, template scripts are generally discouraged. The best practice for using a custom script is to embed it in the process definition at first, and move it to a separate file with its own command line interface once as the code matures.

Reviewer: ```suggestion
Due to the caveats listed above, template scripts are generally discouraged. The best practice for using a custom script is to embed it in the process definition at first and move it to a separate file with its own command line interface once the code matures.
```

:::{note}
- Shell script definitions require the use of single-quote `'` delimited strings. When using double-quote `"` delimited strings, dollar variables are interpreted as Nextflow variables as usual. See {ref}`string-interpolation`.
- Variables prefixed with `!` must always be enclosed in curly brackets, i.e. `!{str}` is a valid variable whereas `!str` is ignored.

> - Shell scripts support the use of the {ref}`process-template` mechanism. The same rules are applied to the variables defined in the template script.

Author: Gonna leave this one alone because shell block is a whole other can of worms
